Logs customization doc: SetResourceBuilder (#3043)
open-telemetry/opentelemetry-dotnet: docs/logs/customizing-the-sdk/README.md

### SetResourceBuilder

> A `ResourceBuilder` provides a `Resource` to the `OpenTelemetryLoggerProvider`.

Reviewer: please borrow as much from https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry#resource.

(need to test to confirm if the ENV variable is applicable to logs as well, and document accordingly)

Author: Do you want the full content copied here or should I link to the doc you shared?

Reviewer: copy here.
(We need to later do some refactoring to extract common concepts like Resource to a common place, and have all signals refer to it.)

Also noticed that, Metrics docs doesn't talk about Resource! Good thing to add it there too.

Author: yeah, i noticed the other docs had TODOs for this content

Author: Done. I had to take out references to TracerProvider. Let me know if it needs additional edits

A [Resource](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md)
is an immutable representation of the entity producing telemetry as
[Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes).

The snippet below demonstrates creating a default `ResourceBuilder` and adding
service metadata.

```csharp
using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder
        .AddOpenTelemetry(options =>
        {
            options.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(
                serviceName: "MyService",
                serviceVersion: "1.0.0"
                ));
            options.AddConsoleExporter();
        });
});

var logger = loggerFactory.CreateLogger<Program>();

logger.LogInformation("Hello Information");
```

> This example generates the following output which includes the customized

Reviewer: ```suggestion
This example generates the following output which includes the Resource information
```

service metadata:

```text
LogRecord.TraceId:            00000000000000000000000000000000
LogRecord.SpanId:             0000000000000000

> LogRecord.Timestamp:          0001-01-01T00:00:00.0000000Z

Reviewer 2: Where is timestamp coming from? It seems wrong.

Author: That was the format for the output, I manually zeroed out the values. 
I contemplated removing most of the example output and only showing the last three lines which show the resource.

Reviewer 2: You can put a valid timestamp like the 1st second of 1970.

Author: done. changed to `1970-01-01T00:00:01.0000000Z`

### SetResourceBuilder

[Resource](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md)
is the immutable representation of the entity producing the telemetry.
If no `Resource` is explicitly configured, the default is to use a resource
indicating this [Telemetry
SDK](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#telemetry-sdk).

> The `SetResourceBuilder` method can be used to set a single `ResourceBuilder`.

Reviewer: ```suggestion
The `SetResourceBuilder` method on `OpenTelemetryLoggerOptions` can be used to set a single `ResourceBuilder`.
```

> The `SetResourceBuilder` method can be used to set a single `ResourceBuilder`.

Reviewer: ```suggestion
The `SetResourceBuilder` method can be used to set a `ResourceBuilder`.
```

If `SetResourceBuilder` is called multiple times, only the last is kept..
It is not possible to change the resource builder *after* creating the

> LoggerFactory.

Reviewer: ```suggestion
`LoggerFactory`.
```

The snippet below shows configuring a custom `ResourceBuilder` to the provider.

```csharp
using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder
        .AddOpenTelemetry(options =>
        {
            options.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(
                serviceName: "MyService",
                serviceVersion: "1.0.0"
                ));
            options.AddConsoleExporter();
        });
});

var logger = loggerFactory.CreateLogger<Program>();

logger.LogInformation("Hello Information");
```

> This example generates the following output which includes the Resource

Reviewer: perhaps not needed to show the actual output here. By this time, users already got started with getting-started, so they should be able to run it and see output themselves.

Author: I'm a fan of including the output. 
I think demonstrating what these settings affect is valuable.

What if we remove the top half of the example and only include the bottom section?
```
Resource associated with LogRecord:
    service.name: MyService
    service.version: 1.0.0
    service.instance.id: 00000000-0000-0000-0000-000000000000
```

Author: I also agree that these are very simple examples and someone should be able to follow along with ease.

Reviewer: I'd suggest to remove actual output. We don't do that anywhere else (except in the getting-started guide). By completing getting-started, we expect user should already know where to look for output.

information:

```text
LogRecord.TraceId:            00000000000000000000000000000000
LogRecord.SpanId:             0000000000000000
LogRecord.Timestamp:          1970-01-01T00:00:01.0000000Z
LogRecord.EventId:            0
LogRecord.EventName:
LogRecord.CategoryName:       Program
LogRecord.LogLevel:           Information
LogRecord.TraceFlags:         None
LogRecord.State:              Hello Information
Resource associated with LogRecord:
    service.name: MyService
    service.version: 1.0.0
    service.instance.id: 00000000-0000-0000-0000-000000000000
```

See [Program.cs](Program.cs) for complete example.

It is also possible to configure the `Resource` by using following
environmental variables:

<!-- markdownlint-disable MD013 -->
| Environment variable       | Description                                        |
| -------------------------- | -------------------------------------------------- |

> | `OTEL_RESOURCE_ATTRIBUTES` | Key-value pairs to be used as resource attributes. See the [Resource SDK specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.5.0/specification/resource/sdk.md#specifying-resource-information-via-an-environment-variable) for more details. |

Reviewer: could you confirm if the ENV variable are respected by LoggingProvider? It should work, but good to add unit tests to it...

Author: added new tests!

Author: I wasn't sure the best place for these, so I created a new file. Please let me know if there's a more appropriate file.

### SetResourceBuilder

> [Resource](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md)

Reviewer: this has references to "sampler" which is a tracing only concept, so need to tweak this accordingly.
Also the user is not building the provider here, instead they build the logger factory, so we need to modify wording accordingly.

Author: I cleaned up this section:

[Resource](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md) is the immutable representation of the entity producing the telemetry.
If no `Resource` is explicitly configured, the default is to use a resource
indicating this [Telemetry SDK](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions#telemetry-sdk).
The `SetResourceBuilder` method can be used to set a single `ResourceBuilder`.
If `SetResourceBuilder` is called multiple times, only the last is kept..
It is not possible to change the resource builder *after* creating the LoggerFactory.
